storage/journal/src/main/java/io/atomix/storage/journal/SegmentedJournalReader.java

```diff
@@ -133,7 +135,7 @@ private void forward(long index) {
   public boolean hasNext() {
     if (!currentReader.hasNext()) {
       JournalSegment<E> nextSegment = journal.getNextSegment(currentSegment.index());
-      if (nextSegment != null) {
+      if (nextSegment != null && nextSegment.index() == getNextIndex()) {
```

Comment:
Avoid inconsistencies in log segments by forcing segments’ last index to align with the next segment’s first index.